API for diploma project

kirill-k88 opened the pull request:
Реализовал API по заданию диплома

kirill-k88 committed:
features: add base project files

kirill-k88 committed:
features: add routes, schemas, controllers

kirill-k88 committed:
features: add correct routes

kirill-k88 committed:
fetures: add all routes and functionality of backend

kirill-k88 committed:
features: changed readme file

kirill-k88 committed:
fix cors url

kirill-k88 committed:
add * to allow all requests sources

kirill-k88 committed:
renamed db to bitfilmsdb

kirill-k88 committed:
add comment

adzhenalieva commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [ ] `GET /movies` — все сохранённые пользователем фильмы;
     _При запросе фильмов, следует в качестве фильтра указать поле owner, чтобы выгрузились только фильмы, сохраненные пользователем. Пример: https://disk.yandex.ru/i/hPiVNFiDueIGjg_
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [ ] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
   - _Путь маршрутов необходимо описать в папке routes. В корневом файле нужно подключить маршрут без описания информации о том, при обращении к какому адресу он должен выполняться. На скриншоте указан допустимый вариант подключения маршрута к приложению: https://snipboard.io/gFMU5T.jpg_
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
     -  _При создании пользователя в схему celebrate к полю name следует добавить .required(), так как это поле обязательное для создания пользователя_
     -   _При сохранении фильма из схемы  валидации celebrate поле owner нужно удалить. Оно не приходит в теле запроса, а будет добавлено в контроллере, куда будет записан id текущего пользователя_
     - _При сохранении фильма в схеме  валидации celebrate у полей не следует указывать мин и макс значения, они не определены заданием. Эти ограничения могут вызвать затруднения при сохранении фильмов на следующих этапах диплома_
  - [ ] Данные валидируются перед добавлением в базу. 
    -  _В схеме фильма следует добавить `required: true` к полю owner, так как оно обязательное при создании фильма_
    - _В схеме  фильма у полей не следует указывать мин и макс значения, они не определены заданием. Эти ограничения могут вызвать затруднения при сохранении фильмов на следующих этапах диплома_
    - _Поля trailerLink, thumbnail и image в схеме фильма следует валидировать. Для валидации можно использовать метод isURL модуля validator: https://www.npmjs.com/package/validator либо проверить регулярным выражением_
    - _Email в модели пользователя также следует валидировать. Для валидации можно использовать метод isEmail модуля validator: https://www.npmjs.com/package/validator_
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [ ] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
     _Сервер недоступен, прошу проверить верно ли указан адрес: https://disk.yandex.ru/i/UY_L_KHKxl9Frg_
- [ ] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [ ] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
     _Смогу засчитать пункт после проверки сервера_
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [ ] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.85--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co или API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 54

## Доп. Комментарии:

kirill-k88 committed:
fix: При запросе фильмов, следует в качестве фильтра указать поле owner, чтобы выгрузились только фильмы, сохраненные пользователем

kirill-k88 committed:
fix: Путь маршрутов необходимо описать в папке routes. В корневом файле нужно подключить маршрут без описания информации о том, при обращении к какому адресу он должен выполняться

kirill-k88 committed:
fix: При создании пользователя в схему celebrate к полю name следует добавить .required(), так как это поле обязательное для создания пользователя
При сохранении фильма из схемы валидации celebrate поле owner нужно удалить. Оно не приходит в теле запроса, а будет добавлено в контроллере, куда будет записан id текущего пользователя
При сохранении фильма в схеме валидации celebrate у полей не следует указывать мин и макс значения, они не определены заданием. Эти ограничения могут вызвать затруднения при сохранении фильмов на следующих этапах диплома

kirill-k88 committed:
fix: В схеме фильма следует добавить required: true к полю owner, так как оно обязательное при создании фильма
В схеме фильма у полей не следует указывать мин и макс значения, они не определены заданием. Эти ограничения могут вызвать затруднения при сохранении фильмов на следующих этапах диплома
Поля trailerLink, thumbnail и image в схеме фильма следует валидировать. Для валидации можно использовать метод isURL модуля validator: https://www.npmjs.com/package/validator либо проверить регулярным выражением
Email в модели пользователя также следует валидировать

kirill-k88 committed:
fix: Сервер недоступен, прошу проверить верно ли указан адре

kirill-k88 committed:
fix: дрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл

kirill-k88 committed:
fix: Конфигурация и константы в отдельных файлах:

kirill-k88 committed:
fix: теперь вместо bearer token использую кукию Добавлен маршрут logout

kirill-k88 committed:
fix: ошибки eslint

adzhenalieva commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
  - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  - [x] Данные валидируются перед добавлением в базу. 
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [ ] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.85--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co или API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 98

## Доп. Комментарии:
